src/indieweb_utils/indieauth/server.py

```diff
@@ -0,0 +1,308 @@
+        :rtype: str
+        :param redirect_uri: The redirect URI of the authorization request.
+        :rtype: str
+        :param decoded_client_id: The decoded client ID of the authorization request.
+        :rtype: str
+        :param decoded_redirect_uri: The decoded redirect URI of the authorization request.
+        :rtype: str
+        :param decoded_expires: The decoded expiration time of the authorization request.
+        :rtype: int
+        :returns: True if the decoded code is valid, False otherwise.
+        :rtype: bool
+    """
+
+    if int(time.time()) > decoded_expires:
+        raise AuthorizationCodeExpiredError("The authorization code has expired.")
+
+    if redirect_uri != decoded_redirect_uri:
+        raise TokenValidationError(
+            "The redirect URI provided does not match the redirect URI in the authorization token."
+        )
+
+    if client_id != decoded_client_id:
+        raise TokenValidationError(
+            "The client ID provided does not match the client ID in the authorization token."
+        )
+
+    return True
+
+
+def generate_auth_token(
+        me: str,
+        client_id: str,
+        redirect_uri: str,
+        response_type: str,
+        state: str,
+        code_challenge: str,
+        code_challenge_method: str,
+        final_scope: str,
+        secret_key: str,
+        **kwargs
+    ) -> str:
+    """
+        Generates an IndieAuth authorization token.
+
+        :param me: The URL of the user's profile.
+        :type me: str
+        :param client_id: The client ID of the authorization request.
+        :type client_id: str
+        :param redirect_uri: The redirect URI of the authorization request.
+        :type redirect_uri: str
+        :param response_type: The response type of the authorization request.
+        :type response_type: str
+        :param state: The state of the authorization request.
+        :type state: str
+        :param code_challenge: The code challenge, used for PKCE.
+        :type code_challenge: str
+        :param code_challenge_method: The code challenge method, used for PKCE.
+        :type code_challenge_method: str
+        :param final_scope: The scopes approved by the user.
+        :type final_scope: str
+        :param secret_key: The secret key used to sign the token.
+        :type secret_key: str
+        :param kwargs: Additional parameters to include in the token.
+        :type kwargs: dict
+        :returns: The authorization token.
+        :rtype: str
+    """
+
+    if not client_id or not redirect_uri or not response_type or (not state and state != ""):
+        raise AuthenticationError("Token request is missing required parameters.")
+
+    if response_type != "code" and response_type != "id":
+        raise AuthenticationError("Only code and id response types are supported.")
+
+    random_string = "".join(random.choice(string.ascii_uppercase + string.digits) for _ in range(10))
```

Comment:
What is random_string used for? I don't recall reading about that in the spec?  I'd suggest we make a function to generate random strings / tokens for us and use that throughout.

```python
import binascii
import os

def generate_token(*, size: int = 20) -> str:
    return binascii.hexlify(os.urandom(size)).decode()

```

Reply:
Good question. That code may be redundant. I added that section per:

> The client first creates a code verifier for each authorization request by generating a random string using the characters [A-Z] / [a-z] / [0-9] / - / . / _ / ~ with a minimum length of 43 characters and maximum length of 128 characters. This value is stored on the client and will be used in the authorization code exchange step later.

This section is described fully in the "[Authorization Request](https://indieauth.spec.indieweb.org/#authorization-request)" part of the spec.

But, it looks like this value is not separate from the code verifier which is already generated separately.

Reply:
After further review, this random string should be code_verifier. In my server-side implementation, I save code_verifier in a session object for later use. I then compare the decoded challenge with the code_verifier during the authorization stage to confirm the integrity of the request. Although that does mean that we should expose the random string in the generate_auth_token function. I shall include that in my next commit.